app/modules/pwm2.c

```diff
@@ -0,0 +1,553 @@
+  {
+    if (isPinSetup(moduleData, i))
+    {
+      updatePinResolutionToInterruptsMultiplier(&moduleData->setupData.pin[i], moduleData->setupData.interruptTimerCPUTicks);
+      updatePinPulseToInterruptsCounter(&moduleData->interruptData.pin[i], &moduleData->setupData.pin[i]);
+      updatePinOffCounter(&moduleData->interruptData.pin[i], &moduleData->setupData.pin[i]);
+    }
+  }
+}
+
+//############################
+// interrupt handler related
+
+static inline void computeIsPinOn(pwm2_pin_interrupt_t *pin, uint16_t *maskOn)
+{
+  if (pin->currentInterruptCounter == pin->pulseInterruptCcounter)
+  {
+    pin->currentInterruptCounter = 1;
+  }
+  else
+  {
+    pin->currentInterruptCounter++;
+  }
+  // ets_printf("curr=%u    on=%u\n", pin->currentInterruptCounter, (pin->currentInterruptCounter < pin->offInterruptCounter));
+  if (pin->currentInterruptCounter < pin->offInterruptCounter)
+  {
+    *maskOn |= pin->gpioMask;
+  }
+}
+
+static inline bool isPinSetup2(const pwm2_interrupt_handler_data_t *data, const uint8_t pin)
+{
+  return data->pin[pin].gpioMask > 0;
+}
+
+static inline uint16_t findAllPinOns(pwm2_interrupt_handler_data_t *data)
+{
+  uint16_t maskOn = 0;
+  for (int i = 1; i < GPIO_PIN_NUM; i++)
+  {
+    if (isPinSetup2(data, i))
+    {
+      computeIsPinOn(&data->pin[i], &maskOn);
+    }
+  }
+  return maskOn;
+}
+
+static inline void setGpioPins(const uint16_t enabledGpioMask, const register uint16_t maskOn)
+{
+  GPIO_REG_WRITE(GPIO_OUT_W1TS_ADDRESS, maskOn);
+  const register uint16_t maskOff = ~maskOn & enabledGpioMask;
+  GPIO_REG_WRITE(GPIO_OUT_W1TC_ADDRESS, maskOff);
+}
+
+static void ICACHE_RAM_ATTR timerInterruptHandler(os_param_t arg)
+{
+  pwm2_interrupt_handler_data_t *data = (pwm2_interrupt_handler_data_t *)arg;
+  setGpioPins(data->enabledGpioMask, findAllPinOns(data));
+}
+
+//############################
+// lua bindings
+
+static int lpwm2_open(lua_State *L)
+{
+  moduleData = os_malloc(sizeof(pwm2_module_data_t));
+  moduleData->interruptData.enabledGpioMask = 0;
+  moduleData->setupData.interruptTimerCPUTicks = 0;
+  moduleData->setupData.interruptTimerTicks = 0;
+  moduleData->setupData.isStarted = false;
+  for (int i = 1; i < GPIO_PIN_NUM; i++)
+  {
+    moduleData->interruptData.pin[i].currentInterruptCounter = 0;
+    moduleData->interruptData.pin[i].gpioMask = 0;
+    moduleData->interruptData.pin[i].offInterruptCounter = 0;
+    moduleData->interruptData.pin[i].pulseInterruptCcounter = 0;
+    moduleData->setupData.pin[i].divisableFrequency = 0;
+    moduleData->setupData.pin[i].duty = 0;
+    moduleData->setupData.pin[i].frequencyDivisor = 0;
+    moduleData->setupData.pin[i].pulseResolutions = 0;
+    moduleData->setupData.pin[i].resolutionCPUTicks = 0;
+    moduleData->setupData.pin[i].resolutionInterruptCounterMultiplier = 0;
+  }
+  return 0;
+}
+
+static int lpwm2_print_setup(lua_State *L)
+{
+  ets_printf("pwm2 : isStarted %u\n", moduleData->setupData.isStarted);
```

Comment:
In general it is a bad idea to include any form of `printf()` in NodeMCU modules, as most production IoTs will not have the UART connected.  Better to return this as an array or even strings so that you leave it to the calling app to print this or return it over the network as, say, a JSON buffer.

Reply:
ok, take a look at latest commit and lets discuss again if not good enough.